Mathlib/Algebra/DirectSum/LinearMap.lean

```diff
@@ -0,0 +1,63 @@
+/-
+Copyright (c) 2023 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import Mathlib.Algebra.DirectSum.Module
+import Mathlib.LinearAlgebra.Trace
+
+/-!
+# Linear maps between direct sums
+
+This file contains results about linear maps which respect direct sum decompositions of their
+domain and codomain.
+
+-/
+
+open Set BigOperators DirectSum
+
+variable {ι R M : Type*} [CommRing R] [AddCommGroup M] [Module R M]
+  {N : ι → Submodule R M} [DecidableEq ι] (h : IsInternal N)
+
+namespace LinearMap
+
+/-- If a linear map `f : M → M₂` respects direct sum decompositions of `M` and `M₂`, then it has a
+block diagonal matrix with respect to bases compatible with the direct sum decompositions. -/
+lemma toMatrix_directSum_collectedBasis_eq_blockDiagonal' [Fintype ι]
+    {M₂ : Type*} [AddCommGroup M₂] [Module R M₂] {N₂ : ι → Submodule R M₂} (h₂ : IsInternal N₂)
+    {κ κ₂ : ι → Type*} [∀ i, Fintype (κ i)] [∀ i, Fintype (κ₂ i)] [∀ i, DecidableEq (κ i)]
+    (b : (i : ι) → Basis (κ i) R (N i)) (b₂ : (i : ι) → Basis (κ₂ i) R (N₂ i))
+    {f : M →ₗ[R] M₂} (hf : ∀ i, MapsTo f (N i) (N₂ i)) :
+    toMatrix (h.collectedBasis b) (h₂.collectedBasis b₂) f =
+    Matrix.blockDiagonal' fun i ↦ toMatrix (b i) (b₂ i) (f.restrict (hf i)) := by
+  ext ⟨i, _⟩ ⟨j, _⟩
+  simp only [toMatrix_apply, Matrix.blockDiagonal'_apply]
+  rcases eq_or_ne i j with rfl | hij
+  · simp [h₂.collectedBasis_repr_of_mem _ (hf _ (Subtype.mem _)), restrict_apply]
+  · simp [hij, h₂.collectedBasis_repr_of_mem_ne _ hij.symm (hf _ (Subtype.mem _))]
+
+variable [∀ i, Module.Finite R (N i)] [∀ i, Module.Free R (N i)]
+
+lemma trace_eq_sum_trace_restrict [Fintype ι]
+    {f : M →ₗ[R] M} (hf : ∀ i, MapsTo f (N i) (N i)) :
+    trace R M f = ∑ i, trace R (N i) (f.restrict (hf i)) := by
+  let b : (i : ι) → Basis _ R (N i) := fun i ↦ Module.Free.chooseBasis R (N i)
+  simp_rw [trace_eq_matrix_trace R (h.collectedBasis b),
+    toMatrix_directSum_collectedBasis_eq_blockDiagonal' h h b b hf, Matrix.trace_blockDiagonal',
+    ← trace_eq_matrix_trace]
+
+lemma trace_eq_sum_trace_restrict' (hN : {i | N i ≠ ⊥}.Finite)
+    {f : M →ₗ[R] M} (hf : ∀ i, MapsTo f (N i) (N i)) :
+    trace R M f = ∑ i in hN.toFinset, trace R (N i) (f.restrict (hf i)) := by
+  set s : Finset ι := hN.toFinset
+  have h_ind : CompleteLattice.Independent (fun i : s ↦ N i) :=
+    h.submodule_independent.comp Subtype.val_injective
+  have h_sup : ⨆ i, N i = ⊤ := h.submodule_iSup_eq_top
+  replace h_sup : ⨆ (i : s), N i = ⊤ := by
+    suffices ⨆ i, ⨆ (_ : N i = ⊥), N i = ⊥ by
+      rw [iSup_split _ (fun i ↦ N i = ⊥), this, bot_sup_eq] at h_sup; simpa [iSup_subtype]
+    simp [iSup_eq_bot]
+  replace h := DirectSum.isInternal_submodule_of_independent_of_iSup_eq_top h_ind h_sup
```

Comment:
I think these results hold for additive monoids if you use the API for `DirectSum.internal` directly rather than going via independence; though I haven't tried.

Reply:
I'm thinking along the lines of:
```
lemma trace_eq_sum_trace_restrict' (hN : {i | N i ≠ ⊥}.Finite)
    {f : M →ₗ[R] M} (hf : ∀ i, MapsTo f (N i) (N i)) :
    trace R M f = ∑ i in hN.toFinset, trace R (N i) (f.restrict (hf i)) := by
  set s : Finset ι := hN.toFinset
  -- TODO: this should be `internal.chooseDecomposition`
  let d' : DirectSum.Decomposition N := {
    decompose' := (Equiv.ofBijective _ h).symm
    left_inv := (Equiv.ofBijective _ h).right_inv
    right_inv := (Equiv.ofBijective _ h).left_inv
  }
  let h : DirectSum.Decomposition (fun i : s => N i) := {
    decompose' := (DirectSum.toModule R _ _ fun i =>
      if hi : i ∈ s then DirectSum.lof R s (N ·) ⟨i, hi⟩ else 0) ∘ₗ (decomposeLinearEquiv N).toLinearMap
    left_inv := sorry
    right_inv := sorry
  }
  ```
  where an extra constructor for `DirectSum.decomposition` would make those sorries trivial

Reply:
That might still be a dead end, but the extra constructors are in #8374

Reply:
Hmm, I'm not sure how hard to fight for this approach. On the one hand it does seem like key results such as:
```lean
   (⨁ (i : { i // N i ≠ ⊥ }), N i) ≃+ ⨁ i, N i
```
should exists somewhere, but on the other, I begin to think the best use of effort would be to generalise the `DirectSum` API so that everything is set up relative to a subset `s : Set ι`.

I need to finish a non-Mathlib piece of work this morning so I'll think about this later today if you feel it's worth pushing for.

Reply:
here's a very messy proof:
```lean
lemma trace_eq_sum_trace_restrict' (hN : {i | N i ≠ ⊥}.Finite)
    {f : M →ₗ[R] M} (hf : ∀ i, MapsTo f (N i) (N i)) :
    trace R M f = ∑ i in hN.toFinset, trace R (N i) (f.restrict (hf i)) := by
  set s : Finset ι := hN.toFinset
  -- TODO: this should be `internal.chooseDecomposition`
  let d' : DirectSum.Decomposition N := h.chooseDecomposition
  let h' : DirectSum.Decomposition (fun i : s => N i) :=
    .ofLinearMap _
      ((DirectSum.toModule R _ _ fun i =>
        if hi : i ∈ s then DirectSum.lof R s (N ·) ⟨i, hi⟩ else 0) ∘ₗ (decomposeLinearEquiv N).toLinearMap)
      (DirectSum.decompose_lhom_ext N fun i => by
        ext x
        dsimp
        simp [←lof_eq_of R]
        split_ifs with h
        · simp [lof_eq_of]
        · suffices x = 0 by simp [this]
          obtain ⟨x, hx⟩ := x
          simp at h
          simpa [h] using hx)
      (by
        ext i x
        -- stupid lof vs of
        simp; simp [lof_eq_of]; simp [←lof_eq_of R]
        split_ifs with h
        · rfl
        · suffices x = 0 by simp [this]
          obtain ⟨x, hx⟩ := x
          simp at h
          simpa [h] using hx)
  rw [← Finset.sum_coe_sort, trace_eq_sum_trace_restrict h'.isInternal (fun i ↦ hf i)]
```

Reply:
I couldn't quite motivate myself to follow this suggestion but as a compromise I did split out the lemma `DirectSum.isInternal_ne_bot_iff` which I hope makes what I'm proposing more palatable.

Reply:
The lemma split is certainly a good idea, thanks! I think probably the best approach overall is:

* Build the `(⨁ (i : { i // N i ≠ ⊥ }), N i) ≃+ ⨁ i, N i` map you describe; I think we probably do have this somewhere
* Define `DirectSum.Decomposition.congrAddEquiv` which takes that equiv and produces `Decomposition N ≃ Decomposition fun i : {i // N i ≠ ⊥ } => N i`
* Define `IsInternal_iff_of_decomposition_equiv`,that turns that result into an `Iff`

The payoff is that `DirectSum.isInternal_ne_bot_iff` becomes true for `AddCommMonoid` (even if that's not helpful for `trace`.

Reply:
I essentially agree with you though I am reluctant to take this detour right now.

I will need this lemma but I have excised it from this pull request in the hopes of unblocking the other changes I am proposing.